pandas/core/reshape.py

```diff
@@ -1250,40 +1290,84 @@ def get_empty_Frame(data, sparse):
         sparse_series = {}
         N = len(data)
         sp_indices = [[] for _ in range(len(dummy_cols))]
-        for ndx, code in enumerate(codes):
-            if code == -1:
-                # Blank entries if not dummy_na and code == -1, #GH4446
-                continue
-            sp_indices[code].append(ndx)
+        for ndx, code, missing in zip(
+                range(len(codes)), codes, missing_codes_msk):
+            if missing:
+                # For missing values, we have to decide what to do.
+                # GH15926
+
+                if dummy_na:
+                    # Then we need a one in the last column.
+                    # GH15926
+                    sp_indices[code].append(ndx)
+
+                if fill_value != 0:
+                    # Then we need to mark these locations to put back another
+                    # fill value later. (Zero fill values will be filled by the
+                    # sparse array implicitly).
+                    # Use a negative index here to code NaNs.
+                    # Offset by -1 to account for zero.
+                    # Have to add to ALL columns, except the
+                    # last one if dummy_na.
+                    # GH15926
+                    if dummy_na:
+                        _num_cols = len(dummy_cols) - 1
+                    else:
+                        _num_cols = len(dummy_cols)
+                    for _code in range(_num_cols):
+                        sp_indices[_code].append(-ndx - 1)
+            else:
+                # Value is not missing so do as normal.
+                # GH15926
+                sp_indices[code].append(ndx)
 
         if drop_first:
             # remove first categorical level to avoid perfect collinearity
             # GH12042
             sp_indices = sp_indices[1:]
             dummy_cols = dummy_cols[1:]
         for col, ixs in zip(dummy_cols, sp_indices):
-            sarr = SparseArray(np.ones(len(ixs), dtype=np.uint8),
-                               sparse_index=IntIndex(N, ixs), fill_value=0,
-                               dtype=np.uint8)
+            sarr = np.ones(len(ixs), dtype=np.float32)
```

Comment:
the dtype selection here is quite tricky actually, you need to use ``pandas.types.infer_dtype_from_scalar`` to figure this out (in concert if there are NaN's). The reason is if ``fill_value=-1`` or any integer, than the dtype needs to be a signed int8, if its 0 it could be uint8, if you have nans then float32 is fine.